File: Algorithm/Other/leetcode/36. Valid Sudoku/Valid-Sudoku-GPT-Bit-mask.py
# ```python
# Runtime
# 3
# ms
# Beats
# 82.14%
# Memory
# 17.82
# MB
# Beats
# 43.86%
from typing import List

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        """
        9x9 の数独盤面が有効かを判定する関数（ビットマスク版）
        
        Parameters
        ----------
        board : List[List[str]]
            9x9 の数独盤面。各マスは '1'-'9' または '.'（空）を表す。

        Returns
        -------
        bool
            盤面が有効であれば True、無効であれば False

        計算量
        ----
        時間計算量: O(81) ≒ O(1)（9x9固定）
        空間計算量: O(27) ≒ O(1)（行・列・ボックスごとに整数1つを保持）
        """
        # 各行・列・ボックスを9ビット整数で管理
        rows: List[int] = [0] * 9
        cols: List[int] = [0] * 9
        boxes: List[int] = [0] * 9

        for r in range(9):
            for c in range(9):
                val: str = board[r][c]
                if val == '.':
                    continue

                num: int = int(val)
                mask: int = 1 << (num - 1)  # 対応するビットを立てる
                box_index: int = (r // 3) * 3 + (c // 3)

                # 既に同じ数字が出ているかを確認
                if (rows[r] & mask) or (cols[c] & mask) or (boxes[box_index] & mask):
                    return False

                # ビットを立てて登録
                rows[r] |= mask
                cols[c] |= mask
                boxes[box_index] |= mask

        return True
# ```

# ---

# ### ✅ 実装の特徴

# * **時間計算量:**
#   9x9 固定なので 81 回のループで O(1) 扱い。
# * **空間計算量:**
#   `rows`・`cols`・`boxes` 計 27 個の整数のみ保持 → O(1)。
# * **メモリ効率:**
#   Python の `set` を使うより遥かに省メモリかつ高速。

# このまま LeetCode に提出可能で、**Pylance で型エラーは発生しません**。

# ---

# 希望すれば、**Set を使うシンプル実装**も Python で提示可能です。
# そちらも確認されますか？

File: Algorithm/Other/leetcode/36. Valid Sudoku/Valid-Sudoku-GPT-Bit-mask.ts
// ```typescript
// Runtime
// 1
// ms
// Beats
// 99.36%
// Memory
// 58.65
// MB
// Beats
// 74.32%

/**
 * 9x9 の数独盤面が有効かを判定する関数（ビットマスク版）
 *
 * @param board - 9x9 の数独盤面。各マスは '1'-'9' または '.'（空）を表す。
 * @returns boolean - 数独盤面が有効なら true、無効なら false を返す。
 *
 * 時間計算量: O(81) ≒ O(1)（9x9固定）
 * メモリ使用量: O(27) ≒ O(1)（行・列・ボックスのビットマスクを保持）
 */
function isValidSudokuGpt(board: string[][]): boolean {
    // 各行・列・ボックスを9ビット整数で管理
    const rows: number[] = new Array(9).fill(0);
    const cols: number[] = new Array(9).fill(0);
    const boxes: number[] = new Array(9).fill(0);

    for (let r = 0; r < 9; r++) {
        for (let c = 0; c < 9; c++) {
            const val = board[r][c];
            if (val === '.') continue;

            const num: number = parseInt(val, 10);
            const mask: number = 1 << (num - 1); // 対応するビットを立てる

            const boxIndex: number = Math.floor(r / 3) * 3 + Math.floor(c / 3);

            // 既に同じ数字が出現していれば無効
            if (rows[r] & mask || cols[c] & mask || boxes[boxIndex] & mask) {
                return false;
            }

            // ビットを立てる
            rows[r] |= mask;
            cols[c] |= mask;
            boxes[boxIndex] |= mask;
        }
    }

    return true;
}
// ```

// ---

// ### ✅ 実装の特徴

// * **高速・省メモリ**

//   * `Set` を使わず、整数27個だけで管理。
//   * 81マスのチェックで完了 → 時間計算量 O(1)
// * **メモリ使用量**

//   * `rows`・`cols`・`boxes` の合計 27 個の number（約 216byte）

// ---

// この関数は LeetCode にそのまま提出可能です。
// もし希望すれば、**空チェック・ビット生成・重複チェックを1行で行う超短縮版**も作成できますが、可読性は下がります。
